src/test/scala/examples/Functional.scala

```diff
@@ -0,0 +1,178 @@
+// This implements a test suite for different rounding modes with all DspTools Ring operations 
+// bku Oct 26 2018
+
+package demo_pkg
+
+import chisel3._
+
+import dsptools.numbers.{Ring, DspReal, DspComplex}
+import dsptools.numbers.implicits._
+
+import chisel3.core.{FixedPoint => FP}
+import dsptools.{DspTester, DspTesterOptions, DspTesterOptionsManager, DspContext, TrimType}
+import dsptools.{NoTrim, RoundHalfUp, StochasticRound}
+
+import org.scalatest.{Matchers, FlatSpec}
+
+// Reference operations
+trait DspArithmetic {
+  
+  val inType  = FP(16.W, 10.BP)
+  val outType = FP(20.W, 12.BP)
+  
+  val complexInType   = DspComplex(inType, inType)
+  val complexOutType  = DspComplex(outType, outType)
+
+  val pipeDepth  = 2
+  val roundType  = StochasticRound
+  
+  // DSP Methods to test
+  def mul[T<:Data with Num[T]]  = (x:T, y:T, z:T) =>  x * y
+  def add[T<:Data with Num[T]]  = (x:T, y:T, z:T) =>  x + y
+  def sub[T<:Data with Num[T]]  = (x:T, y:T, z:T) =>  x - y
+  def div[T<:Data with Num[T]]  = (x:T, y:T, z:T) =>  x / y
+  def mod[T<:Data with Num[T]]  = (x:T, y:T, z:T) =>  x % y
+  def madd[T<:Data with Num[T]] = (x:T, y:T, z:T) =>  x * y + z
+  def msub[T<:Data with Num[T]] = (x:T, y:T, z:T) =>  x * y - z
+
+}
+
+// Generic functional class, parametrized with different combinational operations
+class Functional[A <: Data, B <: Data] (inType:A, outType:B, pipes:Int, rnd:TrimType, op:(A,A,A) => A) extends Module {
+
+  val io = IO (new Bundle {
+    val a   = Input(inType)
+    val b   = Input(inType)
+    val c   = Input(inType)
+    val res = Output(outType)
+  })
+
+  val mres  = Reg(outType)
+ 
+  // TBD: Add Dsp Pipelining tests if required
+  DspContext.alter(DspContext.current.copy(trimType = rnd)) { 
+    mres := op(io.a, io.b, io.c) 
+  }
+ 
+  io.res := mres
+
+}
+
+// Test setup
+trait TestCommon {
+
+  val totalTests = 2
+  
+  val rand  = scala.util.Random
+  val range = 10
+  
+  var x = 0.0
+  var y = 0.0
+  var z = 0.0
+
+}
+
+// Test top
+class FunctionalTester[A <: Data, B <: Data](c: Functional[A,B], op:String) extends DspTester(c) with TestCommon with DspArithmetic {
```

Comment:
I actually can't implement it straightforward as 
```scala
class FunctionalTester[A <: Data, B <: Data](c: Functional[A,B], op:(A,A,A)=>A) extends DspTester(c) with TestCommon with DspArithmetic {...}
```
Since I get a matching warning:
patterns after a variable pattern cannot match (SLS 8.1.1)
[warn] If you intended to match against method mul in trait DspArithmetic, you must use backticks, like: case `mul` =>
And adding backticks yield in a compile error of course

Reply:
DspTester uses Real Numbers on input and output. I wanna use FP and Complex types. So I leave those translations for DspTester impl, but need to use pattern matching to implement a predictor in Reals. If you know how to do better, pls explain